### PR TITLE
Implement stringIsJson method

### PR DESCRIPTION
**Pull request by ipetkovic**
This is the WIP implementation of stringIsJson method.
Could you have a look and see if this makes sense please?

Some remarks:

- I am using custom small boolean stack to keep information about nested structure (current position inside array or inside an object). The requirement was to make this zero-allocation method which I believe is true here but I am using custom small sized boolean stack implemented with plain Long value to avoid heap allocation. Info I found on the net was contradictory on whether array can be allocated on stack or not. I tried few things but failed to make it stack allocated. In the end I decided for a Long value where bits represent state - 0 (inside array), 1 (inside object) which could be an overkill just to avoid heap allocation but I left it anyways :D The downside is that I cannot check strings which have nested depth larger than 58. One option would be to use preallocated array or perhaps not to track this info at all (then we could have false positives for JSON string)
- Number check is not strict - I am just checking for allowed characters there. Can easily change this as well. I was not sure how strict of a check we want

I am happy to hear any comments.

### CODE REVIEW

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (justcoon) [inline core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala:23]: should not be `else` here ?

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (justcoon) [inline core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala:9]: maybe those 3 defs should be in companion object as val

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (justcoon) [inline core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala:28]: this branch creating SmallBooleanStack twice (second time in increaseIdx)

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (justcoon) [inline core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala:14]: setState is private, but name is confusing as it look like mutating function ...

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (justcoon) [inline core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala:22]: maybe 
```scala
object SmallBooleanStack {
    private val undefinedValue = 0xffffffffffffffffL
    
    val undefined: SmallBooleanStack = new SmallBooleanStack(undefinedValue)
}

```

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (justcoon) [inline core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala:34]: maybe 
``` scala
val empty: SmallBooleanStack = new SmallBooleanStack(0L)
```

File: core/shared/src/main/scala/zio/logging/internal/json/JsonValidator.scala
Comment (justcoon) [inline core/shared/src/main/scala/zio/logging/internal/json/JsonValidator.scala:82]: add check 
```if (startIdx < string.length)```

**Comment by justcoon** [comment]
@ipetkovic thank you for contribution

I added some comments, maybe it is good to add some tests, 

we will need them anyway, also even if this impl. will change, tests should not a lot

**Comment by ipetkovic** [comment]
@justcoon Thanks for comments, will update it today probably, also with tests. I already have some but need to add some more cases.
Can you just check remarks in my original comments and let me know what you think?

**Comment by justcoon** [comment]
@ipetkovic 
personally I think, in this case it is about the ratio between performance and correctness,
if we can do very fast "partial" validation in compare to completely correct validation (but much slower one), i think it should be ok in this case

maybe it could be good to check (if you did not do it already) how some others json libraries doing it, or do "validation" benchmark - this implementation vs. some json library implementation,
i did not seen some latest benchmarks in relation to json java or scala libraries, but some time ago I seen that jsoniter was mentioned like fast one 

also considering that this will be internal impl. we can do some improvements in next releases ....

File: core/shared/src/main/scala/zio/logging/internal/json/JsonValidator.scala
Comment (jdegoes) [inline core/shared/src/main/scala/zio/logging/internal/json/JsonValidator.scala:17]: The trick to eliminate use of `SmallBooleanStack` is to _recurse_: this allows you to store on the JVM stack the information on the block type. "Popping" the stack corresponds to exiting one of the recursions, and bubbling up to the next higher-level; "pushing" a boolean on the stack corresponds to recursing.

In order to implement this optimization, you can generalize `checkJson` to take the current state and the boolean. 

Note that when completed, this means we cannot validate JSON of arbitrary depth. We could solve that by switching over to a slower-based validator after we recurse `n` times (where `n` > 200, say). But I'd say solving that is not essential for v1.

**Comment by jdegoes** [comment]
@ipetkovic Overall, great work coming up with this small, efficient validator! I made one suggestion to eliminate the heap usage.

The other thing we need hear is a `Gen`-based test. We need to generate random JSON and verify that this will validate all JSON correctly, even weird JSON with non-ASCII characters, etc.

File: core/shared/src/main/scala/zio/logging/internal/json/JsonValidator.scala
Comment (ipetkovic) [reply core/shared/src/main/scala/zio/logging/internal/json/JsonValidator.scala]: Done

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (ipetkovic) [reply core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala]: Not relevant anymore.

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (ipetkovic) [reply core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala]: Not relevant anymore.

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (ipetkovic) [reply core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala]: Not relevant anymore.

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (ipetkovic) [reply core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala]: Not relevant anymore.

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (ipetkovic) [reply core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala]: Not relevant anymore.

File: core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala
Comment (ipetkovic) [reply core/shared/src/main/scala/zio/logging/internal/json/SmallBooleanStack.scala]: Not relevant anymore.

File: core/shared/src/main/scala/zio/logging/internal/json/JsonValidator.scala
Comment (ipetkovic) [reply core/shared/src/main/scala/zio/logging/internal/json/JsonValidator.scala]: Done.

**Comment by ipetkovic** [comment]
Thank you both for the comments! I updated the implementation to be more efficient using recursive approach.
@jdegoes  Thanks for suggestion for recursion , seemed so obvious after you mentioned it :D
@justcoon it should be strict now. Also, more strict check for numbers makes it even more efficient than my last simple solution. It kind of makes sense since checking for all allowed characters in number with lots of digits is more expensive than checking for digit only for example etc.

Gen-based tests will come soon.

**Comment by ipetkovic** [comment]
@justcoon @jdegoes I did some benchmarking and improved an algorithm a bit. I attached test for JsonValidator and included the fix in LogAppender.

Compared to my first version, this one handles whitespaces much better now. It was a bit naive to use isWhitespace method from RichChar. Now I introduced lookup array for whitespace which improved performance for up to factor of 2.
I noticed this issue with pretty printed JSON files.

I measured times to process 15 different JSON files (500 bytes for up to 700KB) and compared it with play and gson parsers. It is still not fair comparison since these parsers do much more but still...

- worst case: JsonValidator 30% faster
- best case: JsonValidator 185% faster
- average: JsonValidator 88% faster
- median: JsonValidator 91% faster

, play is pretty slow:

- worst case: JsonValidator 100% faster
- best case: JsonValidator 830% faster
- average : JsonValidator 275% faster

I tried to use jsoniter but failed to make it work for an arbitrary JSON structure. Did not spend too much time there but I can in the next days if you like.

Benchmark was done using -XX:+UseEpsilonGC, disabled GC, + all memory preallocated.

There is room to further optimize the algorithm but the biggest issue is, in my opinion, the fact that I am working with String, and using charAt to access characters. Using underlying byte array would make it much more efficient. But this would then result in memory allocation since, as far as I know, it is not possible to get access to internal array. At least not in non-kinky way.

Please let me know what you think.

File: core/shared/src/main/scala/zio/logging/internal/JsonValidator.scala
Comment (justcoon) [inline core/shared/src/main/scala/zio/logging/internal/JsonValidator.scala:7]: maybe 

stringIsJson -> isJson

**Comment by justcoon** [comment]
@ipetkovic 

can you add tests/spec to JsonLogFormatSpec, which covering this feature?

File: core/shared/src/main/scala/zio/logging/internal/JsonValidator.scala
Comment (ipetkovic) [reply core/shared/src/main/scala/zio/logging/internal/JsonValidator.scala]: Done

**Comment by ipetkovic** [comment]
@justcoon I updated with one test. Do you think I should add more? Other tests are covering non-json part so I figured one should do it.

**Comment by justcoon** [comment]
@ipetkovic , i recommend to add some spec for annotations with json value

also it will be good to update `ConsoleJsonApp` in  examples/core  (to use annotation with json value like it is in #643) or add another app, which show that json structured logging is working 
and then also update docs with that example 

thank you

**Comment by ipetkovic** [comment]
@justcoon I updated JsonLogFormatSpec with few more test cases.
Example and docu are also updated.

**Review by justcoon — APPROVED** [review]

**Comment by justcoon** [comment]
@ipetkovic - thank you for contribution

**Comment by ipetkovic** [comment]
@justcoon No problem! It was fun :D
Can you please approve it once more? I squashed everything into one commit.

**Review by justcoon — APPROVED** [review]

**Comment by ipetkovic** [comment]
@justcoon can you merge it? it seems I don't have write access

**Review by jdegoes — APPROVED** [review]

**Comment by jdegoes** [comment]
@ipetkovic Thanks for your work on this! Great job making it 0 allocation. 💪

**Pull request merged**
